### PR TITLE
throws when annotations are not an array full of strings.

### PR DESCRIPTION
**Pull request by wesboyt**
*Issue #, if available:*
#340

*Description of changes:*
Checks that inputs for the function setAnnotations() is an array, and that each annotation within is a string while being written.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/IonTextWriter.ts [hunk 1]
```diff
@@ -328,6 +328,9 @@ export class TextWriter extends AbstractWriter {
     protected writeAnnotations() : void {
         if (this._annotations === null || this._annotations === undefined) return;
         for (let annotation of this._annotations) {
+            if (typeof annotation !== 'string') {
```
Comment (raganhan) [inline src/IonTextWriter.ts]: don't you need to check that `_annotations` is an array?

I believe it'd be better to do this check where `_annotations` array is being mutated. This way you fail fast when it's initialized by anything that's not an array or when trying to  add an element that's not a string.

File: src/AbstractWriter.ts [hunk 2]
```diff
@@ -27,7 +27,13 @@ export abstract class AbstractWriter implements Writer {
     }
 
     setAnnotations(annotations: string[]): void {
-        this._annotations = annotations;
+        if (annotations === undefined || annotations === null) {
+            this._annotations = [];
+        } else if(!Array.isArray(annotations)){
+            throw new Error("Annotations must be an array of strings.")
+        } else {
+            this._annotations = annotations;
```
Comment (raganhan) [inline src/AbstractWriter.ts]: don't you need to check the elements type as well? You are checking it in `IonBinaryWriter#encodeAnnotations`

File: src/IonTextWriter.ts (hunk 1 above)
Comment (wesboyt) [reply src/IonTextWriter.ts]: in the setAnnotations call yes we do.

File: src/AbstractWriter.ts (hunk 2 above)
Comment (wesboyt) [reply src/AbstractWriter.ts]: The elements are getting checked on write, the array is being checked when being set.
users are able to call writer.addAnnotation(string)

Comment (raganhan) [reply src/AbstractWriter.ts]: It's always good to fail fast so it's easier for users to debug and find their mistake

Comment (wesboyt) [reply src/AbstractWriter.ts]: done

File: src/util.ts [hunk 3]
```diff
@@ -27,3 +27,17 @@ export function _hasValue(v: any): boolean {
     return v !== undefined && v !== null;
 }
 
+export function _validateAnnotations(input : string[]) : boolean {
```
Comment (pbcornell) [inline src/util.ts]: please move to AbstractWriter

Comment (wesboyt) [reply src/util.ts]: done

File: src/IonBinaryWriter.ts [hunk 4]
```diff
@@ -311,10 +311,9 @@ export class BinaryWriter extends AbstractWriter {
   }
 
   private encodeAnnotations(annotations: string[]) : Uint8Array {
-    if (!annotations || annotations.length === 0) {
+    if (annotations.length === 0) {
```
Comment (raganhan) [inline src/IonBinaryWriter.ts]: Since this is a private method I'm assuming the invariant here is that annotations is never `null` nor `undefined`, is this correct? 

If it is add a comment about it

Comment (wesboyt) [reply src/IonBinaryWriter.ts]: yes, inputs are enforced or set internally.

**Comment by wesboyt** [comment]
> You are only testing `IonBinaryWriter#setAnnotations`, please test all affect APIs:
> 
> * `IonBinaryWriter#addAnnotations`
> * `IonBinaryWriter#setAnnotations`
> * `IonTextWriter#addAnnotations`
> * `IonTextWriter#setAnnotations`
> 
> Also add tests for:
> 
> * Valid case
> * `null`
> * `undefined`
> * invalid types other than `int`

good call

File: tests/unit/IonTextWriterTest.js [hunk 5]
```diff
@@ -311,6 +311,36 @@
...
+      badWriterTest('Cannot add array of chars.',
+          (writer) => { writer.addAnnotation(['t', 'a', 'c', 'o']), writer.writeInt(5) });
+      badWriterTest('Cannot pass annotations array without a string.',
+          (writer) => { writer.setAnnotations([5]), writer.writeInt(5) });
+      badWriterTest('Cannot pass annotations array containing a non string value.',
+          (writer) => { writer.setAnnotations(['a', 5,'t']), writer.writeInt(5) });
+      badWriterTest('Cannot pass annotations array containing undefined.',
+          (writer) => { writer.setAnnotations([undefined]), writer.writeInt(5) });
+      badWriterTest('Cannot pass annotations array containing null',
+          (writer) => { writer.setAnnotations([null]), writer.writeInt(5) });
```
Comment (raganhan) [inline tests/unit/IonTextWriterTest.js]: add a happy case test for `setAnnotations(null)`

File: tests/unit/IonBinaryWriterTest.js [hunk 6]
```diff
@@ -813,6 +813,22 @@ define([
...
+    errorTest('Cannot add int as annotation.',
+          (writer) => { writer.addAnnotation(5), writer.writeInt(5) });
+    errorTest('Cannot add array of chars.',
+          (writer) => { writer.addAnnotation(['t', 'a', 'c', 'o']), writer.writeInt(5) });
+    errorTest('Cannot pass annotations array containing a non string value.',
+        (writer) => { writer.setAnnotations(['a', 5,'t']), writer.writeInt(5) });
+    errorTest('Cannot pass annotations array containing undefined.',
+        (writer) => { writer.setAnnotations([undefined]), writer.writeInt(5) });
+    errorTest('Cannot pass annotations array containing null',
+        (writer) => { writer.setAnnotations([null]), writer.writeInt(5) });
```
Comment (raganhan) [inline tests/unit/IonBinaryWriterTest.js]: add a happy case test for `setAnnotations(null)`

File: tests/unit/IonTextWriterTest.js (hunk 5 above)
Comment (wesboyt) [reply tests/unit/IonTextWriterTest.js]: done

File: tests/unit/IonBinaryWriterTest.js (hunk 6 above)
Comment (wesboyt) [reply tests/unit/IonBinaryWriterTest.js]: done

File: tests/unit/IonTextWriterTest.js [hunk 7]
```diff
@@ -311,8 +311,44 @@
...
+          (writer) => {
+              writer.stepIn(ion.IonTypes.LIST);
+              writer.stepOut();
+              writer.stepOut();
+          });
+
 
+    // Symbols
+    writerTest('Writes null symbol by detecting null and allows user to pass null into set annotations',
+          (writer) => { writer.setAnnotations(null); writer.writeSymbol(null) },
```
Comment (raganhan) [inline tests/unit/IonTextWriterTest.js]: move it to the annotations section

File: tests/unit/IonBinaryWriterTest.js [hunk 8]
```diff
@@ -625,10 +625,12 @@ define([
       ]);
 
     // Symbols
-
     writerTest('Writes null symbol by detecting null',
       (writer) => { writer.writeSymbol(null) },
         [0x7f]);
+    writerTest('Writes null symbol by detecting null and allows user to pass null into set annotations',
```
[the remark this answers is not in the document]
Comment (wesboyt) [reply tests/unit/IonBinaryWriterTest.js]: nope

File: tests/unit/IonTextWriterTest.js (hunk 7 above)
Comment (wesboyt) [reply tests/unit/IonTextWriterTest.js]: okay

File: tests/unit/IonBinaryWriterTest.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (raganhan) [reply tests/unit/IonBinaryWriterTest.js]: If you disagree with a PR comment explain why or if need clarification say so. Answering "nope" is extremely unhelpful

[the remark this answers is not in the document]
Comment (wesboyt) [reply tests/unit/IonBinaryWriterTest.js]: sorry i meant were gonna throw on null instead as the apis no longer return null.

[the remark this answers is not in the document]
Comment (raganhan) [reply tests/unit/IonBinaryWriterTest.js]: ok. In any case the comment is to use a test description that describes the expected behavior and not what the test is doing. 

When you change to throw on `null` update the description with the expected behavior

[the remark this answers is not in the document]
Comment (wesboyt) [reply tests/unit/IonBinaryWriterTest.js]: gotcha, also the reason i dont plan on getting into parameterization is that were going to delete all these tests for mocha shortly.

[the remark this answers is not in the document]
Comment (raganhan) [reply tests/unit/IonBinaryWriterTest.js]: Ok that makes sense. As we talked offline if you can write the mocha tests now might as well do it, but since the test framework migration is in flight I'm ok waiting until mocha is fully integrated to migrate and reduce the duplication 

I'll let you decide what's the best path here

**Comment by wesboyt** [comment]
tracking further progress in https://github.com/amzn/ion-js/issues/458
